### PR TITLE
Fix backend handling of global floatx

### LINKED ISSUE #3891: set_floatx does not work properly
Once keras and the backend is imported, it is not possible to change the float type using 'set_floatx()'.

So running the following code snippet:

``` python
import keras
print(keras.backend.floatx())
keras.backend.set_floatx('float16')
print(keras.backend.floatx())

# create dummy variable as internally used for weights etc.
a = keras.backend.variable([10])
print(a.dtype)
```

will result in:

```
float32
float16
float32
```

(only tested with theano backend, but with tensorflow backend it should be the same)

Is this the desired behaviour and 'set_floatx()' only for internal use?
Otherwise the backends should request the current float type by calling 'floatx()' instead of using a default value which is set once.

### PR DESCRIPTION
**Pull request by jphalip**
Make sure that changes to the global floatx are effectively taken into account by the backend.
Fixes #3891.

### CODE REVIEW

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -1612,7 +1623,7 @@ def _preprocess_deconv_output_shape(shape, dim_ordering):
 
 
 def _preprocess_conv2d_input(x, dim_ordering):
-    if _FLOATX == 'float64':
+    if floatx() == 'float64':
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: In this case the better solution would to check `dtype(x)`.

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -1624,7 +1635,7 @@ def _preprocess_conv2d_input(x, dim_ordering):
 
 
 def _preprocess_conv3d_input(x, dim_ordering):
-    if _FLOATX == 'float64':
+    if floatx() == 'float64':
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: In this case the better solution would to check `dtype(x)`.

File: keras/backend/tensorflow_backend.py [hunk 3]
```diff
@@ -1636,7 +1647,7 @@ def _preprocess_conv3d_input(x, dim_ordering):
 
 
 def _preprocess_conv2d_kernel(kernel, dim_ordering):
-    if _FLOATX == 'float64':
+    if floatx() == 'float64':
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: In this case the better solution would to check `dtype(x)`.

File: keras/backend/tensorflow_backend.py [hunk 4]
```diff
@@ -1648,7 +1659,7 @@ def _preprocess_conv2d_kernel(kernel, dim_ordering):
 
 
 def _preprocess_conv3d_kernel(kernel, dim_ordering):
-    if _FLOATX == 'float64':
+    if floatx() == 'float64':
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: In this case the better solution would to check `dtype(x)`.

File: keras/backend/tensorflow_backend.py [hunk 5]
```diff
@@ -1697,13 +1708,13 @@ def conv1d(x, kernel, stride=1, border_mode='valid',
         border_mode: string, "same" or "valid".
     '''
     # pre-process dtype
-    if _FLOATX == 'float64':
+    if floatx() == 'float64':
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: In this case the better solution would to check `dtype(x)`.

**Comment by jphalip** [comment]
Thanks @fchollet. I've updated the commit to address your comments and rebased it on master.

**Comment by fchollet** [comment]
LGTM, thanks

**Pull request merged**